versiondb/client/restore_app_db.go

```diff
@@ -0,0 +1,332 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"cosmossdk.io/errors"
+	"github.com/alitto/pond"
+	"github.com/cosmos/iavl"
+	gogotypes "github.com/gogo/protobuf/types"
+	"github.com/linxGnu/grocksdb"
+	"github.com/spf13/cobra"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+
+	"github.com/crypto-org-chain/cronos/cmd/cronosd/open_db"
+	"github.com/crypto-org-chain/cronos/versiondb/extsort"
+	"github.com/crypto-org-chain/cronos/versiondb/memiavl"
+)
+
+const (
+	int64Size = 8
+
+	storeKeyPrefix   = "s/k:%s/"
+	latestVersionKey = "s/latest"
+	commitInfoKeyFmt = "s/%d" // s/<version>
+
+	// We creates the temporary sst files in the target database to make sure the file renaming is cheap in ingestion
+	// part.
+	StoreSSTFileName = "tmp-%s-%d.sst"
+
+	PipelineBufferSize         = 1024
+	DefaultSorterChunkSizeIAVL = 64 * 1024 * 1024
+)
+
+var (
+	nodeKeyFormat = iavl.NewKeyFormat('n', memiavl.SizeHash) // n<hash>
+	rootKeyFormat = iavl.NewKeyFormat('r', int64Size)        // r<version>
+)
+
+func RestoreAppDBCmd(stores []string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "restore-app-db snapshot-dir application.db",
+		Short: "Restore `application.db` from memiavl snapshots",
+		Args:  cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			sstFileSizeTarget, err := cmd.Flags().GetUint64(flagSSTFileSize)
+			if err != nil {
+				return err
+			}
+			sorterChunkSize, err := cmd.Flags().GetUint64(flagSorterChunkSize)
+			if err != nil {
+				return err
+			}
+			concurrency, err := cmd.Flags().GetInt(flagConcurrency)
+			if err != nil {
+				return err
+			}
+
+			snapshotDir := args[0]
+			iavlDir := args[1]
+			if err := os.MkdirAll(iavlDir, os.ModePerm); err != nil {
+				return err
+			}
+
+			// load the snapshots and compute commit info first
+			var lastestVersion int64
+			storeInfos := []storetypes.StoreInfo{
+				// https://github.com/cosmos/cosmos-sdk/issues/14916
+				storetypes.StoreInfo{capabilitytypes.MemStoreKey, storetypes.CommitID{}},
+			}
+			snapshots := make([]*memiavl.Snapshot, len(stores))
+			for i, store := range stores {
+				path := filepath.Join(snapshotDir, store)
+				snapshot, err := memiavl.OpenSnapshot(path)
+				if err != nil {
+					return errors.Wrapf(err, "open snapshot fail: %s", path)
+				}
+				snapshots[i] = snapshot
+
+				tree := memiavl.NewFromSnapshot(snapshot)
+				commitId := lastCommitID(tree)
+				storeInfos = append(storeInfos, storetypes.StoreInfo{
+					Name:     store,
+					CommitId: commitId,
+				})
+
+				if commitId.Version > lastestVersion {
+					lastestVersion = commitId.Version
+				}
+			}
```

Comment:
> in line 85, we might need to close the snapshot if the openSnapshot error happens in the middle.

I think not necessary to complicate the code here, since it's a CLI, after it return, the whole process finished.